### PR TITLE
MSC1730: Mechanism for redirecting to an alternative server during login

### PR DESCRIPTION
**Pull request by richvdh**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/rav/proposal/cs_api_in_login/proposals/1730-cs-api-in-login-response.md)

### CODE REVIEW

**Comment by maxidor** [comment]
fwiw, the full disclosure behind this proposal is that it is used for a closed-source project, possibly paid for, and this MSC reflects that effort, not the #1097 effort. Context is [here](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$15429802808412yNNSE:kamax.io)

**Comment by illyohs** [comment]
What's wrong with a closed source project trying to contribute to the ecosystem and why is it necessary for a disclosure for that?

File: proposals/1730-cs-api-in-login-response.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+It may therefore be useful to provide a mechanism to tell clients which
+endpoint they should use for the client-server (C-S) API after login.
+
+## Proposal
+
+The response to `POST /_matrix/client/r0/login` currently includes the fields
+`user_id`, `access_token`, `device_id`, and the deprecated `home_server`.
+
+We should add to this an optional field `base_cs_url`, which gives a base URL
```
Comment (ara4n) [inline proposals/1730-cs-api-in-login-response.md]: what's the rationale for not using a `homeserver` field and then resolving it to base_cs_url via .well-known? it seems a bit redundant

**Comment by ara4n** [comment]
Yup, the use case for this proposal happened to come from a (probably) closed-source project, although honestly afaik the licensing hasn't been decided yet. But the feature is useful for everyone, hence proposing incorporating it into the core spec, as well as providing a possible solution to #1097.

Currently the expectation on MSCs is that the proposal should stand on its own merits, rather than getting hung up on the licensing of the software which is driving the proposal.  I've made a note (as part of the ongoing #1318 work) to investigate whether we should require disclosures in future, given the risk of covert commercial motives influencing the spec.  Needless to say, the concern is entirely pedantic and academic in the context of this MSC...

**Comment by maxidor** [comment]
> rather than getting hung up on the licensing of the software which is driving the proposal

It's not, as part of the context I linked to. I've summarized what are the actual issues [here](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$15429944038596ZfhlQ:kamax.io).

> Needless to say, the concern is entirely pedantic and academic in the context of this MSC...

As per the context I gave above, it isn't. See a summary of the actual issues [here](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$15429944038596ZfhlQ:kamax.io).
But for the record:
> The problem is not that it's useful or not, or from where it comes from. The problem is that your goal is not to address the protocol issue you picked. Your goal is to address your closed-source project issue. That means:
> 1. There is an assessment the protocol issue is relevant or not, which we cannot make or evaluate ourselves as you do not disclose your project, and it might be wrong in the first place or not relevant (like right now)
> 2. There are "hidden" requirements which make you discard valid input from the community if you stick to the protocol issue you picked, creating tension
> 3. People are mislead into thinking they are working on solving a protocol issue, which is actually not true
> 4. We help you do work you are paid for without knowing it
> 5. You do not disclose from where this comes from. Right now, this is a big conflict of interest, one you did not disclose on your own

To spell it another way:
1. The presentation of the effort mislead contributors into thinking this was to solve a specific issue of the protocol as part of #1097 but instead it was [to solve another problem](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542992382232PdkpK:sw1v.org).
2. It is not possible to offer good and insightful feedback/review if one is not aware of [the full set of requirements](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542989771217OWQFx:sw1v.org) and just have them offer solutions that have no chance to be accepted anyway.
3. It is dishonest to bait people interested into solving #1097 to solve another problem, decide to [not care for the MSC](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542990094223eeegX:sw1v.org) and consider it as not relevant once the real requirements are exposed, and then make them think that [maybe more work will be done on it](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542990330227nxNcP:sw1v.org).
4. and 5. The current situation regarding NV, The non-profit and modular.im is quite difficult as we have no idea from where things are coming, what is behind them - there is basically no transparency.

At this point 2) is the most important: Without the full requirements, it is **not** possible to give valid and insightful feedback and is a waste of time for everyone. The other points are just morally problematic, but that's for you to decide if you care or not. I know I've been tricked here and spent time reviewing work you get paid for that is not aimed to solving community issues but only your own.

**Comment by ara4n** [comment]
> Your goal is to address your closed-source project issue

this is false; you are putting words into our mouth.

the goal of this MSC is to solve the problem of redirecting to an alternative server during login, as per the title. it is expressly trying to solve #1097.

however, it's a valid point that MSCs should be careful about disclosing the original motivation for a feature for full visibility, and we can wrap it into the MSC guidelines. however, the lack of disclosure in #1194 or #1730 is oversight rather than maliciousness, as fun as conspiracy theories are.

**Comment by maxidor** [comment]
> this is false; you are putting words into our mouth.

I am not. @richvdh listed two more requirements that are **NOT** part of the whole #1097 idea. Again, [here they are](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542989771217OWQFx:sw1v.org) and again, once given, the conclusion was that #1730 was possibly never a good choice in the first place ([source](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542989897220baWae:sw1v.org)).

The goal of the MSC was to address a set of requirements for your portal project, not to solve #1097 on its own. #1097 is just a mean to and end, not the end. Else why would @richvdh suddenly drop it and try to solve it another way?

> is oversight rather than maliciousness

Nobody ever claimed otherwise. Personally, The reason behind is not as important as the failure to give full disclosure at all levels so nobody is wasting their time and can also make an informed decision if contributing is in their best interest or not.

**Comment by illyohs** [comment]
Honestly I don't think "disclosure" is the issue or maybe it's just a poor choice of words. Maybe an explicit   Motivation/Real World Usecase section to the template to explain why it's needed would solve the problem of the problem of "disclosure" but I don't see that fixing the "I thought it would solve my issue and issue Y but it doesn't" problem.

**Comment by maxidor** [comment]
> I don't see that fixing the "I thought it would solve my issue and issue Y but it doesn't" problem.

@illyohs Disclosure about the technical requirements that actually need to be addressed, not about the project its for (even tho it greatly helps to know what is the technical background). The proposal is perfectly fine, there is no reason to drop it, yet it was dropped because of undisclosed technical requirements.

In those two technical requirements, only one actually is. For the record, here they are ([source](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542989771217OWQFx:sw1v.org)):
1. I would like the target HS to be able to record the right IP address for the client (is a technical requirement)
2. Implementing HTTP proxying in my portal turns out to be a PITA (is not a technical requirement)

For the first one, I replied to use `X-Forwarded-For` which is the standard header in case of proxying HTTP requests. Yet, it was ignored - at this point I'm guessing because it's actually irrelevant. There are other requirements that are yet undisclosed, which is directly linked to the project being closed-source or even talked about publicly.

So to your point: Full disclosure (in this case technical) would have directly shown this MSC was inappropriate and also that it doesn't seem related to #1097 given how quick it was dropped to work on another approach altogether. It would have avoided wasting people's time to review it, or even consider it in the first place, as I have stated multiple times in my previous comments here and in [#matrix-spec](https://matrix.to/#/#matrix-spec:matrix.org).

**Comment by illyohs** [comment]
I read the discussion again and i see nothing said that that can't be applied to a open source project. I see creating a free VS closed source "disclosure" as something used to generate FUD "creating tension". We should let MSCs stand on technical merit alone and not what licenses people use.

File: proposals/1730-cs-api-in-login-response.md [hunk 2]
```diff
@@ -29,6 +29,35 @@ A representative sequence diagram is shown below.
 
 ![Sequence diagram](images/1730-seq-diagram.svg)
 
+### Potential issues
+
+A significant problem with the proposed architecture is that the portal server
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: The whole MSC seems to be based on "the portal server", yet it is not explained what it is, its purpose, its scope, its requirements. How is anyone not in the confidence supposed to give any kind of meaningful feedback/review on this?

**Comment by richvdh** [comment]
Please can we take the argument about our motivations for proposing spec changes elsewhere. This MSC, like any other, is intended to stand on its own merits; if it does not, or requirements are insufficiently described, that is a bug in the MSC. I have updated the MSC so that it covers all of the points I consider relevant.

File: proposals/1730-cs-api-in-login-response.md (hunk 1 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: I think you're describing exactly what is described under "Perform a .well-known lookup after login" ?

File: proposals/1730-cs-api-in-login-response.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  login.
+
+  This adds latency and overhead, and complicates client implementations.
+
+* It complicates deployment, since each target server has to support a
+  `.well-known` lookup.
+
+* Since the portal already has knowledge of the location of the C-S API for the
+  target homeserver, and has mapped the login request onto the correct HS, it
+  feels redundant to have a separate mechanism which repeats that mapping.
```
Comment (Kegsay) [inline proposals/1730-cs-api-in-login-response.md]: The portal may not know the public-facing URL e.g. in cases where a portal is talking to an internal IP on a private network. It would suck to have to have the portal "know" the base URL for all children HSes.

File: proposals/1730-cs-api-in-login-response.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+* Clients are currently required to do a `.well-known` lookup *before* login,
+  so that they can find the correct endpoint for the `/login` API. That means
+  they will have to do *two* `.well-known` lookups - one before and one after
+  login.
+
+  This adds latency and overhead, and complicates client implementations.
+
+* It complicates deployment, since each target server has to support a
+  `.well-known` lookup.
```
Comment (Kegsay) [inline proposals/1730-cs-api-in-login-response.md]: Ultimately the client needs to know what the base CS URL is, by *some* means. Last I heard, albeit now over a year ago, it was still formed by using the domain of the user ID being logged in (with appropriate overrides if necessary).

I think **all** deployments should be supporting a `.well-known` lookup to formalise what the base CS URL should be. The fact that this was never done isn't an excuse to not implement it IMO.

File: proposals/1730-cs-api-in-login-response.md (hunk 1 above)
Comment (Kegsay) [reply proposals/1730-cs-api-in-login-response.md]: +1 for using `homeserver` field then resolving via `.well-known`. It definitely seems like the clean obvious choice. Sure it's an extra round trip to grab the base CS URL but so what? There's already a large number of requests to extract login flows and submit them...

**Comment by richvdh** [comment]
@Kegsay: yay, thanks for the review \o/. I'm not sure that I agree with all of your technical points, and will follow up later.

However, *please* let's not reopen the discussion about motivations for the MSC - at least not here. Please let's keep the discussion focused on the technical aspects.

**Comment by maxidor** [comment]
@Kegsay I got pinged from #1097 too, spent my time reviewing it, told @richvdh the proposal is perfectly fine (the first commit) and exactly on point. Only to then be told this was NOT about #1097. Did you read the relevant context and the links I gave? It would appear you did not.

@richvdh said the following: ([source](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$1542989897220baWae:sw1v.org))
> anyway now you know my requirements, and they mean that the MSC1730 approach isn't ideal for me

> so yes MSC1730 fulfils the requirement of adding a field to the login response, but actually it turns out that there are other ways to skin this cat

This proposal is not about #1097, it's about that portal for which we are not told the requirements and therefore we cannot review the MSC *for its intended purpose*. How can there be other ways if the MSC is to "officialise" a working process?

I'll say this once again: This proposal is fine. I know it's fine because it's the proposal version of a mechanism already used in other projects which are in production and it works because it's been tested. But this MSC is not about this.

Actually, if you want the full technical picture to make a relevant technical review of this, you also need #1721 and #1731 which give a more complete picture, but still not the full picture. Other pieces of information that are in #matrix-spec but not documented in the various MSCs, quoting @richvdh 
- I would like the target HS to be able to record the right IP address for the client
- Implementing HTTP proxying in my portal turns out to be a PITA
- well, it's a thing that picks a target homeserver based on who is trying to log in which I thought was fairly clear from the description and the sequence diagram, though I concede it is never spelt out explicitly
- oh, it's doing a saml login flow
- [Talking about the CAS feature of Synapse] it's an SSO solution but it still doesn't provide a mechanism to pick an HS

These are directly relevant to the technical review of this. I'm sure there are others but **I cannot check for myself since the code is closed-source**. I've offered several conter-arguments in the room where further discussion took place that fits for the intended goal of #1097 but do not apparently fit for the goal of this portal.

So to sum it up: **three** MSCs are directly related to this portal which is not explained, presented and its technical requirements or features explained. Can we stop the misdirection towards the licensing and actually focus on the technical matters?

**Comment by Kegsay** [comment]
> Did you read the relevant context and the links I gave? It would appear you did not.

I took care to look at every single matrix.to link you sent.

> we are not told the requirements and therefore we cannot review the MSC for its intended purpose

I don't care what their requirements are. I care about what the proposal is, and if it makes sense for the protocol. The idea of giving a base CS URL was always the intention of the protocol so this MSC stands on its own two feet.

> These are directly relevant to the technical review of this. I'm sure there are others but I cannot check for myself since the code is closed-source. 

They *aren't relevant*. If you *work for them* then of course you'd care about the business requirements, but if you don't then you just care about how this affects the protocol.

I am under **absolutely no** illusion that the core team has a lot on their plate trying to keep things alive and people happy. Provided they have a clear separation for spec work vs other work (which https://matrix.org/blog/2018/10/29/introducing-the-matrix-org-foundation-part-1-of-2/ goes a long way to do) then I have no qualms with them proposing things relevant to them because it undergoes the same process as any other person or group working on the protocol.

Now that's out of the way, I won't raise this again in this issue since I appreciate it's inappropriate and only adds to the noise. @maxidor if you want to take this elsewhere then ping me or PM me: @kegan:matrix.org

**Comment by maxidor** [comment]
@Kegsay: if that is true, why was it deprioritized by its author in favor of #1731?

File: proposals/1730-cs-api-in-login-response.md (hunk 1 above)
Comment (turt2live) [reply proposals/1730-cs-api-in-login-response.md]: Honestly I think it's best to avoid the round trip and just return the base URL. It's reasonable to expect that the scenario not be about a portal server, but instead trying to load balance the client-server API on a capable infrastructure. In that scenario, a login server could pick a "homeserver" (web server running whatever proxy/service/matrix-capable-thing) to help distribute clients. In such an environment where one is considering such load balancing, it'd be nice to not have even more round trips to complete.

Comment (Kegsay) [reply proposals/1730-cs-api-in-login-response.md]: I don't understand what the point is you're trying to make. Load balancing can be done using either solution (base URL / .well-known).

The additional round trip isn't ideal, but I'd rather get my data from the destination server than proxy it through another. The former is much more flexible.

Comment (turt2live) [reply proposals/1730-cs-api-in-login-response.md]: That was me trying to understand the benefits of the proposal without the confusing portal server. In the real world, I wouldn't recommend using a portal server and instead recommend the use of a login server, which has a slightly different trust model.

Either way, if the .well-known response is really needed then it may be best to shove it in the response in place of the base url.

Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: > That was me trying to understand the benefits of the proposal without the confusing portal server. In the real world, I wouldn't recommend using a portal server and instead recommend the use of a login server, which has a slightly different trust model.

I'm afraid I don't understand what you see as the difference between a login server and a portal server (which may be partly my fault for not clarifying what I meant by "portal server" - now done, I hope).

Either way, I'm afraid I still don't see what mandating a post-login `.well-known` lookup (as well as the pre-login one we already mandate) gives us above just putting that information in the login response. Which I *think* may be what @turt2live means by

> Either way, if the .well-known response is really needed then it may be best to shove it in the response in place of the base url.

(if that's not what you mean, please clarify)

File: proposals/1730-cs-api-in-login-response.md (hunk 3 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: This is an excellent point. I've removed this paragraph. It was always weak anyway.

File: proposals/1730-cs-api-in-login-response.md (hunk 4 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: Fair. I've updated this.

**Comment by richvdh** [comment]
OK, I have updated the MSC to hopefully be a little clearer about how I think this could be useful.

I think it could be useful in general, not just for my current project. As such I do not believe more details of my current project are at all relevant in the MSC document, though I have answered plenty of questions about it in the `#matrix-spec:matrix.org` room, and a summary is:

>  I have a project which requires users to log in via SAML to a portal server; based on the results of the SAML authentication, the user should automatically be logged into one of several homeservers.
> 
> The homeservers in this project are not in the same datacenter, so X-Forwarded-For is not reliable for proxying without additional authentication between the proxy and the target - which I would prefer to avoid.

**Comment by turt2live** [comment]
@mscbot concern The conversation at https://github.com/matrix-org/matrix-doc/pull/1730#discussion_r236016803 is unresolved.

File: proposals/1730-cs-api-in-login-response.md (hunk 1 above)
Comment (ara4n) [reply proposals/1730-cs-api-in-login-response.md]: I don't care massively strongly on this (and i'm failing to follow the thread due to all the terminology confusion over login & portal servers, which i'm going to assume is a red herring), but for the record, my concerns with `base_cs_url` are:

 * it results in two sources of truth for what the CS URL of a server is: the result of this /login flow, and the value in its .well_known URI.
 * it means you don't get the info from the rest of the .well-known URI metadata, so you may have to go hit it anyway to find out (e.g.) what IS URL to use.

The only benefit i can see is to save a roundtrip to do a .well-known URI lookup, which seems like a premature optimisation.

That said, if I'm missing something, then don't let me block it.

Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: Right, so it seems like Matthew had been missing the fact that the current `.well-known` lookup has to happen before login, so we would effectively be mandating *two* lookups.

> it results in two sources of truth for what the CS URL of a server is: the result of this /login flow, and the value in its .well_known URI.

This is kinda the point - we want to switch CS URL between the initial login, and subsequent interaction.

Whilst I remain fairly strongly opposed to doing a second `.well-known` lookup, it has been suggested by @turt2live and now @ara4n that we could put the equivalent of the well-known data in the `/login` response. I don't think I care very strongly either way about this, but if people think that would be a better solution than a `base_cs_url` field, let me know your reasons here, please.

Comment (turt2live) [reply proposals/1730-cs-api-in-login-response.md]: Having the .well-known object in the response helps reduce the number of requests made by the client, and gives more flexibility for the portal to redirect the client elsewhere. For instance, in a future where the integrations manager is configured by .well-known, it'd be reasonable for a portal to want to point the client elsewhere.

The portal would also have enough information to be able to supply this information, either by requesting it on behalf of the client and proxying the response or by generating the response itself. 

Clients would easily be able to take the object and recofnigure themselves to use it given they'd already have the logic necessary to parse it in the first place. They can probably skip the validation checks at this point, and may wish to optionally alert the user that different variables are being used.

**Comment by maxidor** [comment]
The proposal doesn't take a stance if `/login` should be proxied or redirected but entered final comment period. It feels like the proposal is still incomplete as it doesn't define one, authoritative mechanism and relies on a `307 REDIRECT` which is not currently valid as per r0.4.0 of the C2S spec.

Could this duality be clarified please? And if redirect is chosen, can we spec how the redirect should actually happen (via headers, via body, both, something else?)

**Comment by turt2live** [comment]
For clarity: this has not entered a final comment period. The proposal for it to enter a final comment period is on the floor, however.

**Comment by maxidor** [comment]
As per current spec process:
> Gather feedback as widely as possible.

> The author should neutrally gather the various viewpoints and get consensus, but this can sometimes be time-consuming (or the author may be biased), in which case an impartial 'shepherd' can be assigned to help guide the proposal through this process instead.

> When a member of the Core Team believes that no new discussion points are being made, they will propose a motion for a final comment period (FCP), along with a disposition of either merge, close or postpone.

I have brought up two new discussion points in #matrix-spec that can fulfill the business requirements of the MSC with alternative implementations using a client-integration SSO-based [[1]](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$154327160210921QjcxM:kamax.io)[[2]](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$154327282210956ZiQFO:kamax.io) and as a new 3PID entity [[3]](https://matrix.to/#/!YkZelGRiqijtzXZODa:matrix.org/$154327336910969eVkkN:kamax.io).

Maybe this MSC needs a neutral sheppard to avoid a MSC to be rushed through because of a for-profit client requirement and marked as final comment period in 31h business hours, given barely no time for people to consider all possible options?

**Comment by maxidor** [comment]
> For clarity: this has not entered a final comment period. The proposal for it to enter a final comment period is on the floor, however.

Why is it labelled as such then?

File: proposals/1730-cs-api-in-login-response.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
+# MSC1730: Mechanism for redirecting to an alternative server during login
+
+Complex homeserver deployments may consist of several homeserver instances,
+where the HS to be used for a user session is determined at login time. The HS
+might be chosen based on any of a number of factors, such as the individual
+user, or a simple round-robin to load-balance.
+
+One solution to this is for users to log in via a "portal server", which
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: How does one come across the portal server base URL to call `/login` in the first place, and why can't that mechanism be simply be used to get the HS base URL instead then perform a SSO-like or 3PID-like login using the portal server as backend for the backend (equivalent to a password provider for synapse)?

File: proposals/1730-cs-api-in-login-response.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+We should add to this a `base_cs_url` field, which SHOULD be returned by
+compliant homeservers, which gives a base URL for the client-server API.
+
+As with
+[.well-known](https://matrix.org/docs/spec/client_server/r0.4.0.html#well-known-uri),
+clients would then add `/_matrix/client/...` to this URL to form valid C-S
+endpoints.
+
+One way that this could be used is that the portal server proxies the `/login`
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: This proposal does not offer a single, definitive way - how is one supposed to decide if this way is valid/wanted/relevant compared to the one below?

File: proposals/1730-cs-api-in-login-response.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+request, and passes it on to the target HS, as shown in the sequence diagram below:
+
+![Sequence diagram](images/1730-seq-diagram.svg)
+
+Alternatively, the portal server could redirect the original `login` request to
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: This proposal does not offer a single, definitive way - how is one supposed to decide if this way is valid/wanted/relevant compared to the one above?

File: proposals/1730-cs-api-in-login-response.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+(Note that the deprecated `home_server` field gives the `server_name` of the
+relevant homeserver, which may be quite different to the location of the C-S
+API, so is not of use here. Further we cannot repurpose it, because (a) this
+might break existing clients; (b) it spells homeserver wrong.)
+
+### Notes on proxying vs redirecting
+
+Proxying the `/login` request as shown in the first sequence diagram above
+leads to the following concerns:
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: What is the purpose of listing these concerns? Is there something expected of implementations regarding those? Are they supposed to be used when choosing one of the two solutions above?

File: proposals/1730-cs-api-in-login-response.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+clients would then add `/_matrix/client/...` to this URL to form valid C-S
+endpoints.
+
+One way that this could be used is that the portal server proxies the `/login`
+request, and passes it on to the target HS, as shown in the sequence diagram below:
+
+![Sequence diagram](images/1730-seq-diagram.svg)
+
+Alternatively, the portal server could redirect the original `login` request to
+the target HS with a `307 Temporary Redirect` response:
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: That is not a valid status code from the current C2S API. What is a valid body/headers for this?

File: proposals/1730-cs-api-in-login-response.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+* The target homeserver sees the request coming from the portal server rather
+  than the client, so that the wrong IP address will be recorded against the
+  user's session. (This might be a problem for, for example, IP locking the
+  session, and might affect the `last_seen_ip` field returned by `GET
+  /_matrix/client/r0/devices`.)
+
+  This can be mitigated to some extent via the use of an `X-Forwarded-For`
+  header, but that then requires the portal server to authenticate itself with
+  the target homeserver in some way.
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: This could also be applied to reverse-proxies in front of the C2S API which is the [recommended setup](https://github.com/matrix-org/synapse/issues/2438) for the reference homeserver implementation, and indirectly any other implementations. Why is it a specific concern for a portal but not for a reverse-proxy?

File: proposals/1730-cs-api-in-login-response.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+* It causes additional complexity in the portal server, which must now be
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: This assumes "the portal" would never require any other outbound requests for any other feature or service. This does not seem a reasonable assumption to make and there is no reason to believe that being the case, or not being the case. Both are equally possible.

File: proposals/1730-cs-api-in-login-response.md (hunk 5 above)
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: "portal server" is directly labeled as a solution, but portal server is not defined and not explained. There is no real-life use cases given, no example. If a "portal server" is a solution to this, could this be expanded?

Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: Can this mechanism be applied without such "portal server"?

File: proposals/1730-cs-api-in-login-response.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+  responsible for making outbound HTTP requests.
+
+* It potentially leads to a privacy leak, since the portal server could snoop
+  on the returned access token. (Given that the portal server must be trusted
+  to some extent in this architecture, it is unclear how much of a concern this
+  really is.)
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: I think it would be safe to say: as much as any kind of reverse proxy sitting on front of the C2S API. Those never raised any concerns so far.

File: proposals/1730-cs-api-in-login-response.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+support for `307` redirects among user-agents may vary
+([RFC2616](https://tools.ietf.org/html/rfc2616#section-10.3.8) said "If the 307
+status code is received in response to a request other than GET or HEAD, the
+user agent MUST NOT automatically redirect the request unless it can be
+confirmed by the user", though this appears to have been dropped by
+[RFC7231](https://tools.ietf.org/html/rfc7231#section-6.4.7) and I am unaware
+of any current browsers which do not follow `307` redirects.)
+
+In any case, this is an implementation decision; portal servers can use
+whichever method best suits their needs.
```
Comment (maxidor) [inline proposals/1730-cs-api-in-login-response.md]: Redirection requires a 307 redirect which is not defined in the C2S API spec for `/login`, so this is actually not up to implementations and requires a spec change for one of them.

File: proposals/1730-cs-api-in-login-response.md (hunk 5 above)
Comment (Half-Shot) [inline proposals/1730-cs-api-in-login-response.md]: I'm wondering if we could solve some of the confusion by adding a 1-3 line explanation of what a portal server does.

File: proposals/1730-cs-api-in-login-response.md (hunk 8 above)
Comment (Half-Shot) [reply proposals/1730-cs-api-in-login-response.md]: I assume this is taking the place of a "Potential issues" section.

File: proposals/1730-cs-api-in-login-response.md (hunk 5 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: Clarified in the latest rewrite.

Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: Have rewritten to avoid the troublesome "portal" term

Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: possibly.

Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: Have rewritten to clarify.

File: proposals/1730-cs-api-in-login-response.md (hunk 6 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: The idea was that both ways would fall naturally out of the proposal. I do not agree that it is a fault for a proposal to consider multiple ways that a feature might be used.

However, as discussed below, there are problems with the 307 approach and I have removed it.

File: proposals/1730-cs-api-in-login-response.md (hunk 8 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: yes, it describes some potential issues that might be seen when implementing the proposal.

File: proposals/1730-cs-api-in-login-response.md (hunk 10 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: because the authentication mechanism used for a reverse-proxy may not work here.

File: proposals/1730-cs-api-in-login-response.md (hunk 1 above)
Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: Having experimented with this, using a `well_known` object seems to work well. There was no great objection to it, so I have switched to that.

**Comment by richvdh** [comment]
Right, I have rewritten this, to clarify the problem being solved, to remove the suggestion to do a 307 redirect, and hopefully to explain why other suggestions have been rejected.

**Comment by ara4n** [comment]
lgtm. can we call fcp on it?

**Comment by richvdh** [comment]
> lgtm. can we call fcp on it?

If you like. Go vote for it here: https://github.com/matrix-org/matrix-doc/pull/1730#issuecomment-441758831

**Comment by turt2live** [comment]
@mscbot resolve The conversation at https://github.com/matrix-org/matrix-doc/pull/1730#discussion_r236016803 is unresolved.

**Comment by dbkr** [comment]
My concern with this is how it splits the role of the homeserver up, so in the example, the user gives their homeserver as `matrix.ac.cdl` but their actual homeserver is, say, `matrix.eng.ac.cdl`, so they are effectively entering something into the homeserver field with is not a homeserver at all. For example:

 * What would happen if a client tried to make calls other than /login on `matrix.ac.cdl`?
 * Where do calls to unauthenticated APIs go? `matrix.ac.cdl` presumably? In which case what results would they return?
 * Also adds quite a big gotcha in that plenty of simpler clients will get along fine not implementing this for the 99% of servers that don't use this, then break in subtle ways when you try to log in to a server that does.

That said, I can't think of a better way of doing this that doesn't involve sweeping changes to the way login and identity work.

File: proposals/1730-cs-api-in-login-response.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+problem that we need to switch homeservers between login and initial sync.
+
+An alternative would be for clients to somehow know that they should go through
+the single-sign-on process *before* choosing a homeserver, and for the
+output of the single-sign-on process to indicate the homeserver to use. This
+would require either substantially customised Matrix clients, or substantial
+modifications to the login flow in Matrix, possibly involving authenticating
+against an identity server. The latter is something which could be considered,
+but the scope of the changes required make it impractical in the short/medium
+term.
```
Comment (dbkr) [inline proposals/1730-cs-api-in-login-response.md]: [Moved to inline comment to allow threading]

My concern with this is how it splits the role of the homeserver up, so in the example, the user gives their homeserver as matrix.ac.cdl but their actual homeserver is, say, matrix.eng.ac.cdl, so they are effectively entering something into the homeserver field with is not a homeserver at all. For example:

What would happen if a client tried to make calls other than /login on matrix.ac.cdl?
Where do calls to unauthenticated APIs go? matrix.ac.cdl presumably? In which case what results would they return?
Also adds quite a big gotcha in that plenty of simpler clients will get along fine not implementing this for the 99% of servers that don't use this, then break in subtle ways when you try to log in to a server that does.
That said, I can't think of a better way of doing this that doesn't involve sweeping changes to the way login and identity work.

Comment (richvdh) [reply proposals/1730-cs-api-in-login-response.md]: > What would happen if a client tried to make calls other than /login on matrix.ac.cdl?

more than likely, they would get a 404. But then, why would they do that?

> Where do calls to unauthenticated APIs go? matrix.ac.cdl presumably? In which case what results would they return?

I think that's the same question as above? Which unauthenticated APIs are you thinking of?

> Also adds quite a big gotcha in that plenty of simpler clients will get along fine not implementing this for the 99% of servers that don't use this, then break in subtle ways when you try to log in to a server that does.

Yes, absolutely. I have no good counter-argument to that.

> That said, I can't think of a better way of doing this that doesn't involve sweeping changes to the way login and identity work.

Well, quite :/.

**Comment by erikjohnston** [comment]
Other than perhaps specifying that the dummy server should 401 all other `/_matrix/*` paths, I think this works fine until we implement UIA for `/login`

@mscbot reviewed

**Comment by anoadragon453** [comment]
@mscbot fcp merge

**Comment by turt2live** [comment]
So I completely screwed up here and skipped the FCP step by accident. I've opened https://github.com/matrix-org/matrix-doc/pull/1748 in its place and started an FCP there, copying over the votes [from above](https://github.com/matrix-org/matrix-doc/pull/1730#issuecomment-441758831). 

Sorry about the screw up :(

**Comment by anoadragon453** [comment]
Well, this is why we have a bot :) Now if only it would cooperate :smile:

**Comment by turt2live** [comment]
That would be amazing.

**Comment by richvdh** [comment]
Right so after all that, this should be in its FCP, right?

**Comment by anoadragon453** [comment]
Merged due to FCP completing in #1748.

**Comment by richvdh** [comment]
This is now implemented in riot-web (https://github.com/matrix-org/matrix-react-sdk/pull/2384) and synapse (https://github.com/matrix-org/synapse/pull/4319)

**Comment by turt2live** [comment]
merged via https://github.com/matrix-org/matrix-doc/pull/1790 🎉
